### DISCUSSION TITLE
Deciding abstraction level of TTS

### DISCUSSION THREAD

**Discussion by erogol** [coqui-ai/TTS#9] [General]
After Kirian's latest [PR]( https://github.com/coqui-ai/TTS/pull/4/files), I just wanted to open this discussion...

**What level of abstraction does it look appropriate for 🐸 TTS with respect to your experience?**

Would you prefer a **high level abstraction** where you define abstract classes (e.g. ```TTSAbstract```, ```VocoderAbstract```), as much as possible to template model and new feature implementations? 

Or keeping things in a **script level** and let the models to be self-contained, even maybe in a single file?

I guess good examples are ESPNet for abstraction and HuggingFace for no-abstraction. 
🐸 TTS stands in the middle so far.

In my opinion:

Abstraction...

[+] keep the library organized.
[+] maintains a certain standard.
[+] reduces the code redundancy.
[+] easier to review PRs.
[+] easier to keep maintaining different models as people come and go.
[-] harder to implement new models and features without knowing the library well. That might refrain people to contribute.
[-] abstraction might be constraining for implementing out-of-box features and models if not set correctly.
[-] harder to read and understand the code since everything might be tightened to everything else.

No-Abstraction

[+] Keep the library easy to implement new models and features, especially for newcomers.
[+] No big design choices, so things can change in the future more easily.
[+] Reading the code easier as files are more self-sufficient.
[-] Harder to keep maintaining different models as people come and go.
[-] Too much redundancy.
[-] harder to maintain a coding standard.
[-] harder to keep organized.
[-] harder to review PRs.

Feel free to share your opinion that would really help for the design of 🐸 TTS

**Comment by erogol** [comment]
ignore (move the thread up again)

**Comment by Sleepwalking** [comment]
Hello! I have experience working with both in the context of TTS.

The abstraction-free self-sufficient approach seemed better for short sprints when you just want to get a PoC working under a tight deadline (and once it works, abandon the ship immediately and build a better one from scratch).
* The real problem I encountered was that a few months into the project, the specs start to differ across different algorithms/tasks/datasets; you can't remember for example which data preprocessing steps are used for which models and digging into the code to locate that part will get more and more painful as the project grows.
* There's also code that's actually redundant but was nevertheless copy-pasted over because "it just feels like I might need this `DilatedGatedResidualDepthwiseConvLayerNormBlock` module someday so I'll leave it here".

I switched predominantly to the high level abstraction approach when the technical debt became too much of a burden. Progress was slow in the first few months because of many rounds of refactoring but then the pace slowly ramped up. It's been almost a year and I still refactor a lot. In fact, I probably spend 30% of my time constantly refactoring but overall the speed and code quality are staying in my comfort zone.

**Reply by erogol** [reply]
This is a great input. Thank you ! 

A couple  of follow-ups...

```...in the context of TTS``` meaning :frog:TTS or just in general TTS. 

And along the same lines. Are these bullet points about :frog: TTS ? 

And if they are about :frog: TTS, can you give an example for 

> The real problem I encountered was that a few months into the project, the specs start to differ across different algorithms/tasks/datasets; you can't remember for example which data preprocessing steps are used for which models and digging into the code to locate that part will get more and more painful as the project grows.

Because every model in the same model class (tts, vocoder etc.) uses the same set of processing steps, but the details might differ due to language and audio differences and these are defined in the config files, so it should be enough to read the config to see the differences without going through the code. 

If it is not like your experience, how do you think we can make it better? 

> I switched predominantly to the high level abstraction approach 

Do you have some ref to check your code ?

**Reply by Sleepwalking** [reply]
Sorry for the confusion. I mean general TTS. I work specifically on singing voice synthesis where I basically rolled my own framework on top of PyTorch. There are a few sub tasks. Although the trend in TTS is going end-to-end, in singing we often want to separately model them for being able to manually tweak the results towards a specific artistic preference.

So there's an F0 model, a duration model, a timbre model and a vocoder model. These are totally different tasks with different input and output format/structure/dimensionality, but sometimes they can share the same architecture (e.g. WaveNet) with slight (and annoying!) differences.

One major difference between mine and Coqui/TTS (would you teach me how to type the frog emoji? I really like it) is that I didn't use configuration files out of the fear that it will bloat the code and complicate the workflow. Often I just want to test out a small change really quick. If it's better than the baseline, keep it; otherwise, `git stash` or leave it in a branch. Now to tell which configuration was used to train which model, I simply store the commit hash along with the model.

While this is probably a rather radical approach, so far it works well in my case. One thing I'm not sure is if it can scale to a large team/community based project. I also apologize for not being able to share the code as it's a private commercial project, although I'm happy to share the experience if that helps.

**Reply by erogol** [reply]
:frog: -> ```:frog:``` 

I just watched your demo. It is really impressive. 

Actually :frog: TTS also started as a personal project so I was exactly doing the same thing you do and sign the project state with the commit hash but when people started to contribute and use :frog: TTS it didn't work well. So that's why I try to find a better design allowing an easier collaboration. This also involves a separate discussion about the config management https://github.com/coqui-ai/TTS/issues/21 you might like to check. 

Happy to hear your experience and feel free to share more as you feel necessary : )
